### PR TITLE
Add access to GameProfile cache, a GameProfile resolver and Server.getUser(UUID)

### PR DESCRIPTION
**Pull request by boformer**
This PR adds a service interface that can be used to lookup player UUIDs by name and player names by UUIDS. The service contacts the Mojang auth servers if no cached profile was found in the local cache.
It also adds a ``UserStorage`` interface to get, create or delete a ``User`` object.

This resolves https://github.com/SpongePowered/SpongeAPI/issues/528

Access to cached offline user data is important to:

* complete or validate names of offline players
* lookup the name of an offline player by an UUID stored by a plugin
* ban offline players

Things to discuss:

**Access to the [Name Changes API](http://wiki.vg/Mojang_API#UUID_-.3E_Name_history)?** 
Mojang provides an API which lists up all names and change dates of a player.

### CODE REVIEW

**Comment by Minecrell** [comment]
Cache isn't really a good name for getting `User` instances. The difference between `GameProfile` and `User` is not only persistence or being cached, they represent two completely different use cases:

- `GameProfile`: The identification of a player (UUID + name), as well as optionally a few properties from Mojang's server (like skin and cape). This is generally not persistent, but Minecraft is maintaining a cache for the game profiles for faster profile lookup (`usercache.json`). This includes not only players that have played on the server, but also name->UUID lookups caused through commands.

- `User`: Persistent data of a player, this contains his location in worlds, as well as inventory contents and all the other persistent data of a `Player` - even though not everything is accessible through SpongeAPI. For that to exist, the player has to join the server once. This gets saved together with the world data, in NBT.

A cache is typically used for performance, where entries (can) expire and where it doesn't matter much if certain entries get lost, because they can be queried again. While you could see `User` as cached data being loaded from the disk, a "cache" would be more appropriate for `GameProfile`s, which will be eventually queried from Mojang servers.

Loading `User` data should never trigger any queries to Mojang's server, but rather rely only on the saved persistent user data. If you would want to get a player's data with his newer name (e.g. because he hasn't joined the server since he changed yet) you would need to query name->UUID using a `GameProfile` first and then load the `User` using the returned UUID.

**Comment by boformer** [comment]
@Minecrell What I primarily want is to access the local username<-->uuid storage (only for players who played on the server before).

What do you think of this proposal?

* UserStorage (maybe part of ``Server``)
  * get ``User`` by UUID
* GameProfileCache
  * get stored ``GameProfile`` by UUID
. * get stored ``GameProfile`` by name
. * get list of all profiles
. * list of matching profiles starting with string
* GameProfileResolver
  * get ``GameProfile`` from auth server by UUID
. * get ``GameProfile`` from auth server by name

**Comment by Minecrell** [comment]
@boformer Basically yes, although I think `getUser` should be actually at the same place as `getPlayer` so it would be probably a good idea to move it.

**Comment by boformer** [comment]
I added a new commit which adds the profile cache and resolver and a few new methods in the server interface.

File: src/main/java/org/spongepowered/api/service/userdata/UserDataCache.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+     * @return A list of cached users
+     */
+    List<User> match(String name);
+
+    /**
+     * Returns a list of all cached users.
+     *
+     * @return A list of cached users
+     */
+    List<User> getAll();
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/userdata/UserDataCache.java]: I would imagine this method could be expensive, especially for a lot of previous player data. Perhaps it could be `List<UUID>` instead?

File: src/main/java/org/spongepowered/api/service/profile/GameProfileCache.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * Returns a collection of all cached profiles.
+     *
+     * <p>This collection may include profiles of players who never joined the
+     * server.</p>
+     *
+     * @return A collection of {@link GameProfile}s
+     */
+    ListenableFuture<Collection<GameProfile>> getAll();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileCache.java]: Why is this a future? The cached game profiles should be available without additional query time

File: src/main/java/org/spongepowered/api/service/profile/GameProfileSource.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ */
+public interface GameProfileSource {
+
+    /**
+     * Looks up a {@link GameProfile} by its unique ID.
+     *
+     * @param uniqueId The unique ID
+     * @return The result of the request
+     */
+    ListenableFuture<Optional<GameProfile>> get(UUID uniqueId);
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileSource.java]: I think we can generally say, fetch the game profile or fail (with an exception), so the Optional is not needed

File: src/main/java/org/spongepowered/api/service/profile/GameProfileCache.java (hunk 2 above)
Comment (boformer) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileCache.java]: It's based on this suggestion: https://github.com/SpongePowered/SpongeAPI/issues/528#issuecomment-94164268

In theory, the cache could also be a database, so both service interfaces extend ``GameProfileSource``, which also utilizes ``Future``s.

**Comment by boformer** [comment]
Some more things to discuss:

**Access to the [Name Changes API](http://wiki.vg/Mojang_API#UUID_-.3E_Name_history)?** 
Mojang provides an API which lists up all names and change dates of a player.

**Server.deleteUser(..) method?** 

**A chained cache --> resolver interface?** 
First check the cache, contact Mojang servers if no data is cached or is expired?

**Comment by ST-DDT** [comment]
> get list of all profiles (boolean playedBefore to exclude lookups)

What kind of lookup? I guess you are not refering to get all available GameProfiles from Mojang.

**Comment by boformer** [comment]
No, from the local profile cache.
The cache may also store profiles of players who never played on the server (a plugin just queried the mojang server, and the result is stored in cache).

For the command autocomplete feature, these profiles should not be included.

The completer could also check if a ``User`` object exists for a profile.

To workflow to get a user by name is just becoming very complicated:
* check if profile is in cache
  * no? Query mojang server 
    * no? --> Profile does not exist
* get User object for profile
  * absent? --> player never played on server
* check if user hasPlayedBefore()
...

**Comment by boformer** [comment]
I thought about this again:

* It would make sense to merge the resolver and the cache. Caching should be handled internally (expiration date)
* We need a function that returns a collection of all Users with a name that starts with a string.

**Comment by boformer** [comment]
* The function should not cause a http request to a Mojang server, but use the last known username.

:arrow_right:  We have to store the last known username of a ``User``.

**Comment by MattBDev** [comment]
Any updates on the status of this PR?

**Comment by boformer** [comment]
@MattBDev I think it is fleshed out now. It basically adds 2 new service interfaces:

``UserStorage`` - To get a ``User`` by UUID or last known name
``GameProfileResolver`` - Search the profile cache or contact Mojang servers to get the current name by UUID (or vice versa)

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * <p>The service may cache the data of a request for faster lookups. Note that
+ * the cached data may not always be up to date.</p>
+ *
+ * <p>The returned {@code ListenableFuture} throws an {@link ExecutionException}
+ * caused by a {@link ProfileNotFoundException} if the profile does not exist or
+ * an {@link IOException} if a network error occured.</p>
+ */
+public interface GameProfileResolver
+{
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: Code formatting is wrong here

File: src/main/java/org/spongepowered/api/Server.java [hunk 5]
```diff
@@ -25,8 +25,10 @@
 package org.spongepowered.api;
 
 import com.google.common.base.Optional;
+
 import org.spongepowered.api.entity.Entity;
 import org.spongepowered.api.entity.player.Player;
+import org.spongepowered.api.entity.player.User;
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/Server.java]: Unused import

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.Collection;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+
+import org.spongepowered.api.GameProfile;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+/**
+ * A service which contacts the authentification servers to get a
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: Spelling. "authentification" => "authentication"

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+     * @param profile The profile
+     * @return The user object
+     */
+    User getOrCreate(GameProfile profile);
+
+    /**
+     * Deletes the data associated with a {@link User}.
+     *
+     * @param profile The profile
+     * @return The user object
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: this is a `void` method, no `@return`. I suspect this was copy + paste from above method.

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+     *
+     * @return A {@link Collection} of users.
+     */
+    Collection<User> getAll();
+
+    /**
+     * Returns a collection of matching {@code User}s whose last known user
+     * names start with the given string (case-insensitive).
+     *
+     * @param name
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: `@param lastKnownName`, also need description of parameter

**Comment by simon816** [comment]
Would this PR be an appropriate place to move the method [`createGameProfile`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/GameRegistry.java#L366) out of `GameRegistry` as I don't think it belongs there?

**Comment by MattBDev** [comment]
@simon816 I think that method may have a use that these new classes won't work (I can't think of any myself though) for although I could be wrong.

**Comment by boformer** [comment]
@simon816 The method in GameRegistry has it's own use cases:

* Import ``User`` data from another server (e.g. a Bukkit server)
* Add fake profiles to a ``StatusPingEvent``

File: src/main/java/org/spongepowered/api/Server.java (hunk 5 above)
Comment (me4502) [inline src/main/java/org/spongepowered/api/Server.java]: All of the modifications in this file are unrelated to this PR.

Comment (MattBDev) [inline src/main/java/org/spongepowered/api/Server.java]: Changes on this file should be removed before pulling.

**Comment by MattBDev** [comment]
I think if all the changes on the Server.java file are removed, this can be pulled.

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+import com.google.common.util.concurrent.ListenableFuture;
+
+/**
+ * A service which contacts the authentication servers to get a
+ * {@link GameProfile} by a given UUID or name.
+ *
+ * <p>The service may cache the data of a request for faster lookups. Note that
+ * the cached data may not always be up to date.</p>
+ *
+ * <p>The returned {@code ListenableFuture} throws an {@link ExecutionException}
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: `{@link ListenableFuture}`

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+    /**
+     * Gets the collection of all stored {@link User}s.
+     *
+     * @return A {@link Collection} of users.
+     */
+    Collection<User> getAll();
+
+    /**
+     * Returns a collection of matching {@code User}s whose last known user
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: `{@link User}`

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+     * Deletes the data associated with a {@link User}.
+     *
+     * @param profile The profile
+     */
+    void delete(GameProfile profile);
+
+    /**
+     * Gets the collection of all stored {@link User}s.
+     *
+     * @return A {@link Collection} of users.
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: Remove the period

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+     *
+     * @param uniqueId The UUID of the user.
+     * @return {@link User} or Optional.absent() if not found
+     */
+    Optional<User> get(String lastKnownName);
+
+    /**
+     * Gets the data of a {@link User} by their profile.
+     *
+     * @param uniqueId The UUID of the user.
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: Remove the period

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    Optional<User> get(UUID uniqueId);
+
+    /**
+     * Gets the data of a {@link User} by their last known user name
+     * (case-insensitive).
+     *
+     * <p>To get the current name of a player, use the
+     * {@link GameProfileResolver} service.</p>
+     *
+     * @param uniqueId The UUID of the user.
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: Remove the period

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+/**
+ * Stores the persistent {@link User} data of a {@link Player}.
+ */
+public interface UserStorage {
+
+    /**
+     * Gets the data of a {@link User} by their unique id.
+     *
+     * @param uniqueId The UUID of the user.
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: Remove the period

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+     * @param name The username
+     * @return The result of the request
+     */
+    ListenableFuture<GameProfile> get(String name, boolean useCache);
+
+    /**
+     * Gets a collection of {@link GameProfile}s by their user names
+     * (case-insensitive).
+     *
+     * @param names The user names
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: Need `@param useCache`

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+     * @param uniqueIds The UUIDs
+     * @param useCache true to perform a cache lookup first
+     * @return The result of the request
+     */
+    ListenableFuture<Collection<GameProfile>> getAllById(Iterable<UUID> uniqueIds, boolean useCache);
+
+    /**
+     * Looks up a {@link GameProfile} by its user name (case-insensitive).
+     *
+     * @param name The username
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: Need `@param useCache`

**Comment by MattBDev** [comment]
@boformer are you able to update the PR with the changes @simon816 suggested

**Comment by boformer** [comment]
@MattBDev Done.

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.service.profile;
+
+import java.io.IOException;
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: Wrong import order, `java` imports should be below others. (See [guidelines](https://docs.spongepowered.org/en/contributing/guidelines.html))

**Comment by boformer** [comment]
@simon816 I finally found out how to control this with eclipse

**Comment by MattBDev** [comment]
:exclamation: *waits impatiently for sponge team to review this*

**Comment by RobertHerhold** [comment]
@MattBDev 
Just chill out for a minute. These guys have full-time jobs, and they are *very* busy at times. They will review this when they get to it.

**Comment by Zidane** [comment]
@SpongePowered/developers 

Final call before this is merged. I need some feedback.

**Comment by me4502** [comment]
looks good to me

**Comment by Minecrell** [comment]
How would you resolve a `GameProfile` **only** from the cache (if available), without querying Mojang's servers? Sometimes it might be not that important so you just want it if available.

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+     * @return The user object
+     */
+    User getOrCreate(GameProfile profile);
+
+    /**
+     * Deletes the data associated with a {@link User}.
+     *
+     * @param profile The profile
+     */
+    void delete(GameProfile profile);
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: What would happen here if the user is currently online?

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+    Collection<User> getAll();
+
+    /**
+     * Returns a collection of matching {@link User}s whose last known user
+     * names start with the given string (case-insensitive).
+     *
+     * @param lastKnownName The user name
+     * @return The result of the request
+     */
+    Collection<User> match(String lastKnownName);
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: Couldn't we provide the same also for `GameProfile`s? There will be most likely more `GameProfile`s cached than `User`s, so you could provide more profiles than users (e.g. when used for tab completion).

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+     * @param profile The profile
+     */
+    void delete(GameProfile profile);
+
+    /**
+     * Gets the collection of all stored {@link User}s.
+     *
+     * @return A {@link Collection} of users
+     */
+    Collection<User> getAll();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: Considering that this can be quite inefficient for bigger uses (which may have 5,000+ players persisted) this should have _at least_ some note that this can be quite inefficient depending on the player count on the server.

Comment (simon816) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: I think it should return a collection of UUIDs as it is much cheaper to load many UUID objects.

Comment (me4502) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: I feel GameProfile would be more useful.

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 18 above)
Comment (MattBDev) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: I think it would make sense to delete inventory and other data if they are online

Comment (simon816) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: It would cause sync issues with the client unless everything is updated on the client. Location, inventory, hunger, health etc. Maybe this should fail (and return a boolean false) if the user is online.

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 19 above)
Comment (boformer) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: The problem is that cached gameprofiles expire (at least in vanilla). Also, not all cached gameprofiles are players who played on the server.

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 18 above)
Comment (boformer) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: Or throw an exception?

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 20 above)
Comment (boformer) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: Ok, I will change it to GameProfile.

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 19 above)
Comment (boformer) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: @Minecrell Do you think this method should also return GameProfiles for performance reasons? At least that would be sufficient for tab completion.

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 18 above)
Comment (MattBDev) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: An exception would work.

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 19 above)
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: If you change the other method this should also return `GameProfile`, yes.

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 18 above)
Comment (ryantheleach) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: The API has largely avoided checked exceptions where possible, why add one here, now?

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 20 above)
Comment (ryantheleach) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: If it's that likely to be inefficient, maybe make some sort of a lazy collection, or provide search methods or some way to page it?

**Comment by Minecrell** [comment]
Can you remove [`GameProfile.getUser()`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/GameProfile.java#L43-L48)? With the additions in here these methods should be preferred over the current one.

**Comment by boformer** [comment]
@minecrell yes, I will. Also thought about that today.

**Comment by boformer** [comment]
I have removed the getUser method from GameProfile and changed the return types of the resource-intensive methods to Collection<GameProfile>. I also changed the return type of the delete method to boolean to indicate the success of the deletion.

I think this is ready.

**Comment by Minecrell** [comment]
I'll repeat this question from above:
> How would you resolve a `GameProfile` **only** from the cache (if available), without querying Mojang's servers? Sometimes it might be not that important so you just want it if available.

**Comment by boformer** [comment]
@Minecrell If you want to lookup a player who played on the server before, you can use getUser(UUID/Name).getProfile(), though that's not very elegant.

Do you think it is necessary to have direct access to the profile cache (which also contains players who did not play on the server)?

I could add a third service for that: ``GameProfileCache``, similar to ``GameProfileResolver`` (maybe without Futures).

Or replace the ``boolean useCache`` parameter with a ``ResolveMethod`` enum parameter: CACHE, REMOTE, CACHE_REMOTE.

What do you prefer?

**Comment by Minecrell** [comment]
I would basically like a way to access the cache _without_ any delay (-> without a future). Sometimes you don't want (or can't) halt the execution of your code to wait for the exectution of a future. IMO there should also be a way to access only the cached profiles with direct access.

I know that you made it that way because of https://github.com/SpongePowered/SpongeAPI/pull/580#discussion_r28693330, however there will be most likely a local memory cache anyway simply because that's much faster than querying a remote server everytime. And if not, this cache could simply never return anything.

**Comment by Minecrell** [comment]
@boformer Actually never mind, I realized there is really not a big use case for it.

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+public interface GameProfileResolver {
+
+    /**
+     * Looks up a {@link GameProfile} by its unique ID.
+     *
+     * @param uniqueId The unique ID
+     * @param useCache true to perform a cache lookup first
+     * @return The result of the request
+     */
+    ListenableFuture<GameProfile> get(UUID uniqueId, boolean useCache);
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: What do you think about separating this in like `get()` and `query()` instead of the parameter,  it makes reading the code more clear: `resolver.query(uuid)` vs `resolver.get(uuid, false)` (you always have to look what exactly was this boolean again)

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * Deletes the data associated with a {@link User}.
+     *
+     * <p>This might not work if the user is logged in.</p>
+     *
+     * @param profile The profile
+     * @return true if the deletion was successful
+     */
+    boolean delete(GameProfile profile);
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: Couldn't this take a user? This would ensure the given parameter is really a user that exists.

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+     *
+     * <p>Note that this method is resource-intensive depending on the amount of
+     * stored data.</p>
+     *
+     * <p>Use {@link #get(GameProfile)} to get the {@link User} data
+     * corresponding to a {@link GameProfile}.</p>
+     *
+     * @return A {@link Collection} of {@link GameProfile}s
+     */
+    Collection<GameProfile> getAll();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: While these both methods return a `GameProfile` already, I think they would also additionally make sense with the data from the cache. For example for a command adding player heads to the inventory you don't really care if the user has played on the server already, you want too provide as many options in the tab complete as possible.

**Comment by Minecrell** [comment]
You have a checkstyle warnings in here https://travis-ci.org/SpongePowered/SpongeAPI/jobs/62137985#L257

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java (hunk 21 above)
Comment (boformer) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: Or just ``get(UUID)`` (Shortcut for ``get(UUID, true)``)and ``get(UUID, boolean)``

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 22 above)
Comment (boformer) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: A ``User`` object could also be deleted (if you call the method twice).

I will just add both methods...

File: src/main/java/org/spongepowered/api/service/user/UserStorage.java (hunk 23 above)
Comment (boformer) [inline src/main/java/org/spongepowered/api/service/user/UserStorage.java]: For that use case, we should add similar methods to the profile resolver service:

``getAllCachedProfiles()`` and ``matchCachedProfile(String)``

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java (hunk 21 above)
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: I prefer get and query.
Because the difference between them is very clear due to the method naming.

Comment (boformer) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: I'm not sure about this. Mojang recommends to cache profiles.
I don't want to provide a shortcut for "contact servers directly".

That's why i prefer 2 methods with the same name. Only developers who really want to force a server lookup will use the boolean function.

Comment (progwml6) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: The cache bypass is not the best thing to let people near

Comment (MattBDev) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: Agreed. It's best to have it default to true because performing too many direct requests to Mojang servers can cause the the server to be blacklisted from accessing the API.
Most developers will most likely prefer to use `get(UUID uniqueId)` which will be a lot safer and avoid this risk.

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    ListenableFuture<Collection<GameProfile>> getAllById(Iterable<UUID> uniqueIds, boolean useCache);
+
+    /**
+     * Returns a collection of matching cached {@link GameProfile}s whose last
+     * known user names start with the given string (case-insensitive).
+     *
+     * <p>This collection may also contain profiles of players who never played
+     * on the server!</p>
+     *
```
Comment (MattBDev) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: If this only fetches from the local cache then how can it contain profiles of players who have never played?

Comment (boformer) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: Whenever you do a lookup with get(UUID/name), the profile is saved in the local cache.

You can also lookup profiles of players who never played on the server.

Comment (MattBDev) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: Thanks for this clarification.

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+     * @return The result of the request
+     */
+    Collection<GameProfile> match(String lastKnownName);
+
+    /**
+     * Gets a collection of all cached {@link GameProfile}s.
+     *
+     * @return A {@link Collection} of {@link GameProfile}s
+     */
+    Collection<GameProfile> getAllCachedProfiles();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: Why not just `getAll`?

**Comment by Minecrell** [comment]
https://travis-ci.org/SpongePowered/SpongeAPI/jobs/62295585#L264

**Comment by boformer** [comment]
I squashed the commits.

**Comment by MattBDev** [comment]
:heart_eyes: Looks awesome

**Comment by boformer** [comment]
I fixed the travis errors and renamed the method to ``getAllCached()``.

File: src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+     * @return The result of the request
+     */
+    ListenableFuture<Collection<GameProfile>> getAllById(Iterable<UUID> uniqueIds, boolean useCache);
+
+    /**
+     * Gets a collection of all cached {@link GameProfile}s.
+     *
+     * @return A {@link Collection} of {@link GameProfile}s
+     */
+    Collection<GameProfile> getAllCached();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/service/profile/GameProfileResolver.java]: I think this should be either `getCachedProfiles()` or `getAll()` but `getAllCached()` sounds a bit weird

**Comment by boformer** [comment]
I renamed the method to ``getAllCachedProfiles`` to make clear what it does. ``getAll`` is misleading.

**Comment by Minecrell** [comment]
@boformer `getAllCachedProfiles` is essentially the same as `getCachedProfiles`, the all is kinda redundant in the name.

**Comment by boformer** [comment]
You are right.

File: src/main/java/org/spongepowered/api/service/profile/ProfileNotFoundException.java [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.service.profile;
+
+/**
+ * Thrown by a profile lookup service if a profile does not exist.
+ */
+public class ProfileNotFoundException extends Exception {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/service/profile/ProfileNotFoundException.java]: Couldn't this extend some other exception than just `Exception`?

Comment (gabizou) [inline src/main/java/org/spongepowered/api/service/profile/ProfileNotFoundException.java]: Never mind. It's fine to just extend `Exception` in this case.

**Comment by simon816** [comment]
Why is this a service? I don't think it should be possible for plugins to replace the implementation of the  `GameProfileResolver` and `UserStorage`.

**Comment by MattBDev** [comment]
I didn't realize that this was being added as a service. Extra plugins just to get a User shouldn't be needed. This should be only implemented by sponge

**Comment by boformer** [comment]
@simon816 First of all, almost any system is a service, even the command and event system.

Maybe it would even make sense to have a custom GameProfileResolver with a custom caching solution etc.

**Comment by modwizcode** [comment]
That may just represent a more fundamental problem with the way services
are being used.

On Thu, May 14, 2015 at 4:23 PM, boformer <notifications@github.com> wrote:

> @simon816 <https://github.com/simon816> First of all, almost any system
> is a service, even the command and event system.
>
> Maybe it would even make sense to have a custom GameProfileResolver with a
> custom caching solution etc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SpongePowered/SpongeAPI/pull/580#issuecomment-102175101>
> .
>



-- 

Starbuck Johnson, Java Developer
www.modwiz.com | starbuck@modwiz.com

**Comment by boformer** [comment]
@modwizcode Services are a part of the Sponge philosophy. There are certain core services that can't be overridden. I don't see a problem with that.

**Comment by modwizcode** [comment]
Alright I guess I misunderstood the goal of services.

On Thu, May 14, 2015 at 4:41 PM, boformer <notifications@github.com> wrote:

> @modwizcode <https://github.com/modwizcode> Services are a part of the
> Sponge philosophy. There are certain core services that can't be
> overridden. I don't see a problem with that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SpongePowered/SpongeAPI/pull/580#issuecomment-102178139>
> .
>



-- 

Starbuck Johnson, Java Developer
www.modwiz.com | starbuck@modwiz.com

**Comment by ryantheleach** [comment]
If it were possible for UserStorage to be overridden it would be awesome.

Store all the users out to a database instead, and you already have cross server inventory compatibility.

**Comment by simon816** [comment]
If plugins were able to implement this, they would need to provide an implementation of `User`, which of course would be a problem as the implementation will be expecting it to be an instance of their own user (say `SpongeUser`), ultimately leading to cast exceptions.

**Comment by MattBDev** [comment]
I think if anyone was to implement this in their plugin, they would do so only if they knew what they were doing. I retract my previous statement of not adding this as a service.
